### PR TITLE
[Merged by Bors] - feat: port RingTheory.Noetherian

### PR DESCRIPTION
**Pull request by mattrobball**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/Noetherian.lean [hunk 1]
```diff
@@ -0,0 +1,663 @@
...
+
+
+open Set
+
+open BigOperators Pointwise
+
+/-- `IsNoetherian R M` is the proposition that `M` is a Noetherian `R`-module,
+implemented as the predicate that all `R`-submodules of `M` are finitely generated.
+-/
+-- Porting name: should this be Noetherian?
```
Comment (jcommelin) [inline Mathlib/RingTheory/Noetherian.lean:71]:
```suggestion
-- Porting note: should this be renamed to `Noetherian`?
```

File: Mathlib/RingTheory/Noetherian.lean [hunk 2]
```diff
@@ -178,82 +179,100 @@ end Module
...
 variable [Module R M] [Module R P]
 
 open IsNoetherian
 
-include R
-
-theorem isNoetherian_of_ker_bot [IsNoetherian R P] (f : M →ₗ[R] P) (hf : f.ker = ⊥) :
-    IsNoetherian R M :=
-  isNoetherian_of_linearEquiv (LinearEquiv.ofInjective f <| LinearMap.ker_eq_bot.mp hf).symm
+-- Porting note: more typeclass search juggling
```
Comment (jcommelin) [inline Mathlib/RingTheory/Noetherian.lean:193]: This goes from a perfectly readable statement to a huge `@`-jungle.
I'm quite worried about what this means for the 800 files that depend on this. Do you know if the original statement works when `etaExperiment` is enabled?

File: Mathlib/RingTheory/Noetherian.lean [hunk 3]
```diff
@@ -369,47 +391,80 @@ variable {R M P : Type _} {N : Type w} [Ring R] [AddCommGroup M] [Module R M] [A
...
+        rw [le_antisymm_iff, (this x y).symm, (this y x).symm, ←le_antisymm_iff, imp_self]
+        trivial⟩,
+      by dsimp [GT.gt]; simp only [lt_iff_le_not_le, (this _ _).symm]; tauto⟩
 #align finite_of_linear_independent finite_of_linearIndependent
 
-/-- If the first and final modules in a short exact sequence are noetherian,
-  then the middle module is also noetherian. -/
+/-- If the first and final modules in a short exact sequence are Noetherian,
+  then the middle module is also Noetherian. -/
+/- Porting note: etaExperiment helps with SemilinearMapClass instance and coercion
```
Comment (jcommelin) [inline Mathlib/RingTheory/Noetherian.lean:418]: Please include the original output of mathport in the porting note. I think we should remind ourselves that this went from ~5 lines to ~30 lines.

File: Mathlib/RingTheory/Noetherian.lean [hunk 4]
```diff
@@ -369,47 +391,80 @@ variable {R M P : Type _} {N : Type w} [Ring R] [AddCommGroup M] [Module R M] [A
...
+        (M →ₗ[R] N) LinearMap.instSemilinearMapClassLinearMap _ f hf
+    · exact @Submodule.giMapComap R R N P _ _ _ _ _ _ (RingHom.id _)
+        (N →ₗ[R] P) LinearMap.instSemilinearMapClassLinearMap g hg _
+    · intro; rw [@Submodule.map_comap_eq R R M N _ _ _ _ _ _ (RingHom.id _)
+        (M →ₗ[R] N) LinearMap.instSemilinearMapClassLinearMap, inf_comm]
+    · intro a; rw [@Submodule.comap_map_eq R R N P _ _ _ _ _ _ (RingHom.id _) RingHomSurjective.ids
+        (N →ₗ[R] P) LinearMap.instSemilinearMapClassLinearMap g a, h]
 #align is_noetherian_of_range_eq_ker isNoetherian_of_range_eq_ker
 
+/- Porting note: this seems to causing a diamond with Ring.toSemiring when going to
```
Comment (jcommelin) [inline Mathlib/RingTheory/Noetherian.lean:454]: Please mention `lean4#2074` in the porting note.

**Comment by mattrobball** [comment]
`set_option synthInstance.etaExperiment true`.  Please only use for testing.

_Originally posted by @gebner in https://github.com/leanprover/lean4/issues/2074#issuecomment-1439279264_

Is this no longer current guidance?

File: Mathlib/RingTheory/Noetherian.lean (hunk 2 above)
Comment (mattrobball) [reply Mathlib/RingTheory/Noetherian.lean]: There are only 2-3 TC searches that occur repeatedly and cause all of this mess. `etaExperiment` fixes those though it did error on one other thing for me. I did some digging into the search belt my last commit but I might be able to pin down the issue more.

**Comment by Ruben-VandeVelde** [comment]
There's a dozen occurrences already in mathlib4, so I assume not

**Comment by mattrobball** [comment]
I ask with no preference. (Turning it on would save time to finish my taxes :) But some more formal guidance on using/documenting it might be helpful.

**Comment by mattrobball** [comment]
I made a [Zulip topic](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/using.20etaExperiment) since I don’t want to waste my time fixing things they don’t need it

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**